support deprecated attribute in type constructors (#3715)
gleam-lang/gleam: CHANGELOG.md

### Compiler

> - The compiler now allows deprecating specific constructors of a custom type using the `@deprecated` attribute:

Reviewer: Let's call them variants

  ```gleam
  // Updated `Number` to `RealNumber`s only
  pub type RealNumber {
    @deprecated("`I` am not a real number")
    I

>     One

Reviewer: Can we give these some other names? `I` is a bit hard to read. Perhaps we can use the name of hash algorithms and deprecate md5?

    // ...
  }

  pub fn to_num(n: RealNumber) -> Int {
    case n {
      I -> calc_i(give: n), // Warning: deprecated value `I`
      One -> 1
      // ..

>     }

Reviewer: Let's show it being used to construct a value rather than pattern matching

### Compiler

> - The compiler now allows the `@deprecated` attribute on custom type constructor records/variants.

Reviewer 2: ```suggestion
- The compiler now allows deprecating specific constructors of a custom type using the `@deprecated` attribute:
```
